spec/controllers/devise/two_factor_authentication_controller_spec.rb

```diff
@@ -1,51 +1,58 @@
 require 'rails_helper'
 
 describe Devise::TwoFactorAuthenticationController, devise: true do
-  describe 'update' do
+  describe '#update' do
     let(:user) { create(:user, :signed_up) }
 
+    context 'when resource is fully authenticated' do
+      it 'redirects to the dashboard' do
+        sign_in user
+        get :new
+
+        expect(response).to redirect_to dashboard_index_path
+      end
+    end
+
```

Comment:
For this test to effectively verify that an OTP code can't be submitted when already signed in, it would need to also verify that `#authenticate_otp` is not called. Simply verifying that the action redirects to the dashboard is not enough because the redirect could also happen even if the update action was allowed to be accessed.

Alternatively, you could break this up into 2 unit tests:
- one that verifies that the controller includes a before_action for `check_already_authenticated`, like here: https://github.com/18F/identity-idp/blob/master/spec/controllers/dashboard_controller_spec.rb
- one that tests the `check_already_authenticated` method, like here:
https://github.com/18F/identity-idp/blob/master/spec/controllers/application_controller_spec.rb#L48-L63

That would be my preferred approach.